extensions/native/recursion/src/fri/two_adic_pcs.rs

```diff
@@ -165,10 +165,36 @@ pub fn verify_two_adic_pcs<C: Config>(
             );
             builder.cycle_tracker_end("verify-batch");
 
+            builder.cycle_tracker_start("cache-generator-powers");
+            // truncate index_bits to log_max_height
+            let index_bits_truncated = index_bits.slice(builder, 0, log_max_height);
+
+            // b = index_bits
+            // w = generator of order 2^log_max_height
+            // we first compute `w ** (b[0] * 2^(log_max_height - 1) + ... + b[log_max_height - 1])` using a square-and-multiply algorithm.
+            let w = config.get_two_adic_generator(builder, log_max_height);
+            let res = builder.exp_bits_big_endian(w, &index_bits_truncated);
+
+            // we now compute:
+            // tag_exp[log_max_height - i] = g * w ** (b[log_max_height - i] * 2^(log_max_height - 1) + ... + b[log_max_height - 1] * 2^(log_max_height - i))
+            // using a square-and-divide algorithm.
+            // g * res is tag_exp[0]
+            let tag_exp: Array<C, Felt<C::F>> = builder.array(log_max_height);
```

Comment:
can you rename or make it more clear `tag_exp` is used below as the shifted generator in a coset